controller/src/main/java/org/jetbrains/MavenController.java

```diff
@@ -1,96 +1,146 @@
 package org.jetbrains;
 
+import ch.epfl.scala.bsp4j.*;
 import org.apache.maven.shared.invoker.*;
-
 import org.jetbrains.maven.server.EventPacket;
+
 import java.io.File;
 import java.io.IOException;
-
+import java.net.ServerSocket;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.Properties;
 
 public class MavenController {
-    private final String projectDirectory;
     Invoker invoker;
 
-    public MavenController(String projectDirectory) {
+    BuildClient client;
+    public MavenController() {
+        this(null);
+    }
+    public MavenController(BuildClient client) {
+        this.client = client;
         invoker = new DefaultInvoker();
-        this.projectDirectory = projectDirectory;
     }
 
-    private void startServer(int port) {
-        Thread messenger = new Thread(() -> {
-            try {
-                EventServer server = new EventServer(port);
-                while (server.alive()) {
-                    EventPacket eventPacket = server.getPacket();
-                    String message =  eventPacket.getEvent();
-                    if(message != null)
-                        System.out.println(eventPacket.getEvent());
-                }
-            } catch (IOException | ClassNotFoundException e) {
-                e.printStackTrace();
-            }
-        });
+    private synchronized Integer startServer(BuildClient client) {
+        int port = 0;
+        try {
+            ServerSocket serverSocket = new ServerSocket(0);
+            port = serverSocket.getLocalPort();
 
-        messenger.start();
+            new Thread(new Runnable() {
+                private ServerSocket serverSocket;
 
-        System.out.println("Event Server Started");
-    }
+                public Runnable init(ServerSocket serverSocket) {
+                    this.serverSocket = serverSocket;
+                    return this;
+                }
 
+                @Override
+                public void run() {
+                    try {
+                        EventServer server = new EventServer(serverSocket);
+
+                        while (server.alive()) {
+                            EventPacket eventPacket = server.getPacket();
+                            String message =  eventPacket.getEvent();
+                            if(message != null) {
+                                TaskProgressParams event = new TaskProgressParams(new TaskId("todo"));
+                                event.setEventTime(System.currentTimeMillis());
+                                event.setMessage(eventPacket.getEvent());
+                                client.onBuildTaskProgress(event);
+                            }
+                        }
+                    } catch (IOException | ClassNotFoundException e) {
+                        e.printStackTrace();
```

Comment:
Let's log this error properly after https://github.com/bsp-capstone/maven-bsp/pull/4 is merged.

Reply:
you mean with Log4j right?

Reply:
Yep